tests/test_buffers.py

```diff
@@ -11,6 +11,32 @@
 np = pytest.importorskip("numpy")
 
 
+@pytest.mark.parametrize(
+    ("cpp_name", "expected_codes"),
+    [
+        ("PyObject *", ["O"]),
+        ("bool", ["?"]),
+        ("std::int8_t", ["b"]),
+        ("std::uint8_t", ["B"]),
+        ("std::int16_t", ["h"]),
+        ("std::uint16_t", ["H"]),
+        ("std::int32_t", ["i"]),
+        ("std::uint32_t", ["I"]),
+        ("std::int64_t", ["q"]),
+        ("std::uint64_t", ["Q"]),
+        ("float", ["f"]),
+        ("double", ["d"]),
+        ("long double", ["g", "d"]),
+        ("std::complex<float>", ["Zf"]),
+        ("std::complex<double>", ["Zd"]),
+        ("std::complex<long double>", ["Zg", "Zd"]),
+        ("", ["UNKNOWN"]),
+    ],
+)
+def test_format_descriptor_format(cpp_name, expected_codes):
+    assert m.format_descriptor_format(cpp_name) in expected_codes
```

Comment:
I would add an assert that the format descriptor is a valid numpy format descriptor

https://numpy.org/doc/stable/reference/generated/numpy.format_parser.html#numpy-format-parser

assert np. format_parser(blah).dtype is not None

Reply:
Done-ish. See my comments in the test code.

What a lucky man I was to never have seen #1908 before.

I see Windows isn't happy about my latest version of the test. Fixing...

Reply:
> I would add an assert that the format descriptor is a valid numpy format descriptor

@Lalaland Thanks a lot for nudging me in that direction!

Using `np. format_parser()` didn't work out, but via d432ce75b3963cc759afa640f669157cf110b291 I got on the right track, and after that taught me what I needed to look for, I quickly found it here:

https://github.com/pybind/pybind11/blob/d72ffb448c58b4ffb08b5ad629bc788646e2d59e/include/pybind11/buffer_info.h#L173-L190

From there it was only a small step to add a public (not in `namespace detail`) interface for the existing functionality (029b157540f18b81bad38c92d666b6414b3ef547), and then your validation idea basically fell into place.

The new public interface is also exactly what I need for a clean solution here:

https://github.com/pybind/pybind11_abseil/blob/a8fed7557747fa33e5a04844738baf5f6a1c8d1b/pybind11_abseil/absl_casters.h#L442